eth/protocols/snap/sync.go

```diff
@@ -0,0 +1,718 @@
+		// If the sync was cancelled, abort
+		select {
+		case <-cancel:
+			return false, nil
+		default:
+		}
+		// Generate a random request ID (clash probability is insignificant)
+		id := uint64(rand.Int63())
+
+		// Track the request and send it to the peer
+		s.lock.Lock()
+		s.storageReqs[peer.id] = &storageRequest{
+			id:     id,
+			root:   stroot,
+			origin: s.nextSlot,
+			cancel: cancel,
+		}
+		s.lock.Unlock()
+
+		if err := peer.RequestStorageRange(id, root, account, s.nextSlot, maxRequestSize); err != nil {
+			return false, err
+		}
+		s.storageRequests++
+
+		// Wait for the reply to arrive
+		res := <-s.storageResps
+		if res == nil {
+			return false, errors.New("unfulfilled request")
+		}
+		s.storageSynced += uint64(len(res.slots))
+		s.storageProofs += uint64(len(res.bounds))
+
+		// Hack
+		if res.nodes == nil {
+			break
+		}
+		// Persist every finalized trie node that's not on the boundary
+		it := res.nodes.NewIterator(nil, nil)
+		for it.Next() {
+			// Boundary nodes are not written, since they are incomplete
+			if _, ok := res.bounds[common.BytesToHash(it.Key())]; ok {
+				continue
+			}
+			// Node not a boundary, persist to disk
+			batch.Put(it.Key(), it.Value())
+			s.bloom.Add(it.Key())
+
+			s.storageNodes++
+			s.storageBytes += common.StorageSize(common.HashLength + len(it.Value()))
+		}
+		it.Release()
+
+		if err := batch.Write(); err != nil {
+			return false, err
+		}
+		batch.Reset()
+
+		// Storage range processed, step to the next chunk
+		log.Debug("Persisted range of storage slots", "account", account, "slot", res.last)
+		s.report(false)
+
+		// If the response contained all the data in one shot (no proofs), there
+		// is no reason to continue the sync, report immediate success.
+		if len(res.bounds) == 0 {
+			return true, nil
+		}
+		s.nextSlot = common.BigToHash(new(big.Int).Add(res.last.Big(), big.NewInt(1)))
+	}
+	return false, nil
+}
+
+// OnAccounts is a callback method to invoke when a range of accounts are
+// received from a remote peer.
+func (s *Syncer) OnAccounts(peer *Peer, id uint64, hashes []common.Hash, accounts [][]byte, proof [][]byte) error {
+	peer.Log().Trace("Delivering range of accounts", "hashes", len(hashes), "accounts", len(accounts), "proofs", len(proof))
+
+	// If the request is stale, discard it
+	s.lock.Lock()
+	req, ok := s.accountReqs[peer.id]
+	if !ok || req.id != id {
+		peer.Log().Warn("Unexpected account range packet")
+		s.lock.Unlock()
+		return nil
+	}
+	delete(s.accountReqs, peer.id)
+	root := s.root
+	s.lock.Unlock()
+
+	// If the response is unavailable snapshot, forward to the requester
+	if len(hashes) == 0 && len(accounts) == 0 && len(proof) == 0 {
+		select {
+		//case <-req.cancel:
+		case s.accountResps <- nil:
+		}
+		return nil
+	}
+	// Reconstruct a partial trie from the response and verify it
+	keys := make([][]byte, len(hashes))
+	for i, key := range hashes {
+		keys[i] = common.CopyBytes(key[:])
+	}
+	nodes := make(light.NodeList, len(proof))
```

Comment:
This is a bit inefficient. Perhaps it doesn't matter for accounts, but it does for storage. If the set is small enough to fit in a a batch, we shouldn't provide the proofs. For small storage tries, that could make quite a big difference. 
This diff fixes it for me

```diff
diff --git a/eth/protocols/snap/sync.go b/eth/protocols/snap/sync.go
index 298dc3b479..4818f63b10 100644
--- a/eth/protocols/snap/sync.go
+++ b/eth/protocols/snap/sync.go
@@ -510,13 +510,16 @@ func (s *Syncer) OnAccounts(peer PeerIF, id uint64, hashes []common.Hash, accoun
        for i, key := range hashes {
                keys[i] = common.CopyBytes(key[:])
        }
-       nodes := make(light.NodeList, len(proof))
-       for i, node := range proof {
-               nodes[i] = node
+       var proofDb ethdb.KeyValueReader
+       if len(proof) > 0{
+               nodes := make(light.NodeList, len(proof))
+               for i, node := range proof {
+                       nodes[i] = node
+               }
+               proofDb = nodes.NodeSet()
        }
-       proofdb := nodes.NodeSet()
 
-       db, tr, err := trie.VerifyRangeProof(root, req.origin[:], keys, accounts, proofdb, proofdb)
+       db, tr, err := trie.VerifyRangeProof(root, req.origin[:], keys, accounts, proofDb, proofDb)
        if err != nil {
                return err
        }
```